spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSchemaEvolutionSuite.scala

```diff
@@ -0,0 +1,318 @@
+import org.apache.spark.sql.types._
+
+
+/**
+ * Suite covering widening columns and fields type as part of automatic schema evolution when the
+ * type widening table feature is supported.
+ */
+class DeltaTypeWideningSchemaEvolutionSuite
+    extends QueryTest
+    with DeltaDMLTestUtils
+    with DeltaSQLCommandTest
+    with DeltaTypeWideningTestMixin
+    with DeltaMergeIntoTypeWideningSchemaEvolutionTests {
+
+  protected override def sparkConf: SparkConf = {
+    super.sparkConf
+      .set(DeltaSQLConf.DELTA_SCHEMA_AUTO_MIGRATE.key, "true")
+  }
+}
+
+/**
+ * Tests covering type widening during schema evolution in MERGE INTO.
+ */
+trait DeltaMergeIntoTypeWideningSchemaEvolutionTests
+    extends MergeIntoSQLTestUtils
+    with MergeIntoSchemaEvolutionMixin
+    with DeltaTypeWideningTestCases {
+  self: QueryTest with DeltaTypeWideningTestMixin with DeltaDMLTestUtils =>
+
+  import testImplicits._
+
+  for {
+    testCase <- supportedTestCases
+  } {
+    test(s"automatic type widening in merge ${testCase.fromType.sql} -> ${testCase.toType.sql}") {
+      withTable("source") {
+        testCase.additionalValuesDF.write.format("delta").saveAsTable("source")
+        append(testCase.initialValuesDF)
+
+        // We mainly want to ensure type widening is correctly applied to the schema. We use a
+        // trivial insert only merge to make it easier to validate results.
+        executeMerge(
+          tgt = s"delta.`$tempPath` t",
+          src = "source",
+          cond = "0 = 1",
+          clauses = insert("*"))
+
+        assert(readDeltaTable(tempPath).schema("value").dataType === testCase.toType)
+        checkAnswer(
+          readDeltaTable(tempPath).select("value").sort("value"),
+          testCase.expectedResult.select($"value".cast(testCase.toType)).sort("value"))
+      }
+    }
+  }
+
+  for {
+    testCase <- unsupportedTestCases
+  } {
+    test(s"unsupported automatic type widening in merge " +
+      s"${testCase.fromType.sql} -> ${testCase.toType.sql}") {
+      withTable("source") {
+        testCase.additionalValuesDF.write.format("delta").saveAsTable("source")
+        append(testCase.initialValuesDF)
+
+        // Test cases for some of the unsupported type changes may overflow while others only have
+        // values that can be implicitly cast to the narrower type - e.g. double ->float.
+        // We set storeAssignmentPolicy to LEGACY to ignore overflows, this test only ensures
+        // that the table schema didn't evolve.
+        withSQLConf(SQLConf.STORE_ASSIGNMENT_POLICY.key -> StoreAssignmentPolicy.LEGACY.toString) {
+          executeMerge(
+            tgt = s"delta.`$tempPath` t",
+            src = "source",
+            cond = "0 = 1",
+            clauses = insert("*"))
+          assert(readDeltaTable(tempPath).schema("value").dataType === testCase.fromType)
+        }
+      }
+    }
+  }
+
+  test("type widening isn't applied in merge when schema evolution is disabled") {
```

Comment:
Do we also have a test for when the feature/table property is disabled but `DELTA_SCHEMA_AUTO_MIGRATE` is set to true?

Reply:
We do in MergeIntoSchemaEvolutionSuite, specifically tests covering storeAssignmentPolicy, including source long to target int.

I added a simple test with type widening disabled here, as a kind of smoke test when only this suite is run and not the full MergeIntoSchemaEvolutionSuite,